Add blacklist functionality

koenaad opened the pull request:
I have added functionality to blacklist crates, as described in _#536 Add support for blacklisting crates_.

### Changes

- add a table `blacklisted_crates` to keep track of crate names
- add a new module `db::blacklist` to interact with `blacklisted_crates`
- expose blacklist functionality in the cratesfyi CLI
- check if a crate is on the blacklist before building its docs and return Err if so

The cratesfyi CLI now looks like this:

```
USAGE:
    cratesfyi [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    blacklist           Blacklist operations
    build               Builds documentation in a chroot environment
    daemon              Starts cratesfyi daemon
    database            Database operations
    help                Prints this message or the help of the given subcommand(s)
    queue               Interactions with the build queue
    start-web-server    Starts web server
```

and

```
USAGE:
    cratesfyi blacklist [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    add       Add a crate to the blacklist
    help      Prints this message or the help of the given subcommand(s)
    remove    Remove a crate from the blacklist
```

### TODO's

 - [x] update the [README](https://github.com/rust-lang/docs.rs/blob/master/README.md#cli)

### Open questions / issues

 - I asked here https://github.com/rust-lang/docs.rs/issues/536#issuecomment-569424625 whether only storing a crate name was enough or if we will need more granularity
 - should we validate crate names that are added to the blacklist?

jyn514 commented:
Thanks for the hard work! Before I read the code, a few preliminary comments:

- I would expect the CLI command to be `cratesfyi database blacklist` to match the `db::blacklist` module

> could there be a case in which only certain versions of a 'bad' crate are blocked and other not?

Yes, but I would expect those to be handled with a variant of `delete-crate`, not by blacklisting the whole crate. After all, how would we know to only blacklist a specific version until that version has been published? See also https://github.com/rust-lang/docs.rs/issues/528

> do we want to be able to specify a version after which builds are stopped? I.e., "all versions of crate xyz after version 1.2.3".

This might be nice in the future but I don't think its necessary in a first pass at this.

> The simple implementation I have working now only stores a crate name in blacklisted_crates and does not build a crate if the name exists in blacklisted_crates. Is this enough? (For now?)

Yes, I think that's a good approach.

> should we validate crate names that are added to the blacklist?

As in make sure a crate with that name exists? That's not a bad idea, but it would need an override button (maybe `--force`) in case the database is wrong/out of date/the crate hasn't been published yet.

> check if a crate is on the blacklist before building its docs and return Err if so

Have you tested this? Currently I think we panic if the build returns an error.

jyn514 commented:
Can you also add a `cratesfyi blacklist show` or `list` subcommand that says which crates are currently blacklisted?

koenaad commented on src/docbuilder/rustwide_builder.rs:
```diff
@@ -242,6 +243,11 @@ impl RustwideBuilder {
         let conn = connect_db()?;
+
+        if is_blacklisted(&conn, name)? {
+            return Err(::failure::err_msg("crate is on the blacklist"));
```
I actually copied this from a few lines above 🙈 https://github.com/rust-lang/docs.rs/blob/master/src/docbuilder/rustwide_builder.rs#L122
It is indeed not necessary here.

This seems to be the result from when this code was moved from `src/utils/build_doc_rustwide.rs` in https://github.com/rust-lang/docs.rs/commit/f23d0c1d884fc7115522ef31b3ab75a2367fb430.

jyn514 replied:
Gotcha, good code archeology lol. I think we can leave fixing the other call for another day.

koenaad replied:
This code has been removed with the last commit. Resolved!

koenaad commented on src/bin/cratesfyi.rs:
```diff
@@ -75,6 +75,21 @@ pub fn main() {
+            .subcommand(SubCommand::with_name("add")
+                .about("Add a crate to the blacklist")
+                .arg(Arg::with_name("CRATE_NAME")
+                    .index(1)
```
Oh great! I agree with dropping `.index(1)`.

But looking into `takes_value`: https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.takes_value

> Specifies that the argument takes a value at run time.

I'm actually not sure what it does... If I remove both `.index(1)` and `.takes_value(true)`, clap still works correctly. So I'm a bit hesitant to add it.

jyn514 replied:
Huh, it looks like `takes_value` is only required for flags, not for positional arguments. Ok in that case we can remove both altogether :)

koenaad commented:
> I would expect the CLI command to be `cratesfyi database blacklist` to match the `db::blacklist` module

This sounds reasonable to me. I found it difficult to determine a 'correct' place for the `blacklist` module. In the original issue #536 @pietroalbini mentioned the command should be `cratesfyi blacklist add <CRATE>`. But the blacklist is something that only lives in the database after all...

I feel like I don't have enough context of how docs.rs is operated to make a good decision here.
But `cratesfyi database blacklist add <CRATE>` seems fine for me. It also makes sense to limit the amount of subcommands under`cratesfyi`.

> Yes, but I would expect those to be handled with a variant of delete-crate, not by blacklisting the whole crate. After all, how would we know to only blacklist a specific version until that version has been published? See also #528

Okay, so blacklist is all about disallowing future versions of a crate to be built. Gotcha.

> > do we want to be able to specify a version after which builds are stopped? I.e., "all versions of crate xyz after version 1.2.3".
>
> This might be nice in the future but I don't think its necessary in a first pass at this.

Agree. I guess this can be followed up by #528?

> As in make sure a crate with that name exists? That's not a bad idea, but it would need an override button (maybe `--force`) in case the database is wrong/out of date/the crate hasn't been published yet.

It is not immediately clear to me how we would go about this, so I propose we skip it for now (so no validation of blacklisted crates in this PR) and document this as an issue if it seems useful enough?

koenaad commented:
> I tested this locally and this causes the builder to panic. Not sure what the proper behavior should be, but we should either return Ok() or handle the error:
>
> ```
> thread 'main' panicked at 'Building documentation failed: ErrorMessage { msg: "crate is on the blacklist" }', src/libcore/result.rs:1165:5
> ```

I am also unsure about the proper behaviour of `build_package`. The meaning of the return type `Result<bool>` is unclear to me.
I assume the boolean indicates whether the build was successful? And an Err is returned when something failed that can be retried?

I've returned an Err as mentioned by @pietroalbini in https://github.com/rust-lang/docs.rs/issues/536#issuecomment-569271377:

> 2. In the build process, use the connection to query the table you just created, and **return an error if the crate we're trying to build is blacklisted**.

After some investigation into how `build_package` is called:

- `RustwideBuilder::build_local_package` and `cratesfyi build crate` will unwrap the Result and cause a panic --> seems fine to me since you are only building this specific crate
- `RustwideBuilder::build_world` will log the error and continue --> fine for me
- `DocBuilder::build_next_queue_package` logs the error but retries building the crate --> **this is not good**

So, I'd guess logging an info message and then returning `Ok(false)` would be more appropriate?

jyn514 commented:
> I feel like I don't have enough context of how docs.rs is operated to make a good decision here.
Operated is a strong word ... it's more like making changes to a live database :laughing: see https://forge.rust-lang.org/infra/docs/docs-rs.html for common things to do with `cratesfyi`.

> > As in make sure a crate with that name exists? That's not a bad idea, but it would need an override button (maybe --force) in case the database is wrong/out of date/the crate hasn't been published yet.

> It is not immediately clear to me how we would go about this

You could check in the `crates` table under the `name` column. That table is updated frequently from the crates.io index. If that seems like a pain though, it could certainly wait until later.

> So, I'd guess logging an info message and then returning Ok(false) would be more appropriate?

Yes, that would be fine.

> DocBuilder::build_next_queue_package logs the error but retries building the crate --> this is not good

Correct, and this is how `build_package` is most commonly called in production as well.

jyn514 commented:
> I assume the boolean indicates whether the build was successful? And an Err is returned when something failed that can be retried?

The boolean indicates whether the crate _should_ be built or not, the Error indicates that an unsuccessful build occurred (and the crate should be built).

dns2utf8 commented on src/db/blacklist.rs:
```diff
@@ -0,0 +1,84 @@
+
+pub fn is_blacklisted(conn: &Connection, name: &str) -> Result<bool, Error> {
+    let rows = conn.query(
+        "SELECT COUNT(*) FROM blacklisted_crates WHERE crate_name = $1;",
```
I would use `SELECT COUNT(1) ...` because the SQL Engine does not have to expand `*` to every column in the table. Granted it is only one in this table

jyn514 replied:
Can you expand on this? My impression was that `COUNT(*)` doesn't depend on any columns and only counts the rows. There is not much information on this in the postgres docs: https://www.postgresql.org/docs/current/functions-aggregate.html

dns2utf8 replied:
Hi @jyn514 
I readup on todays implementations and I have to retract. 
Todays databases appear to have an optimization for `COUNT(*)` if there are no `NULL` values in the table.
Since you have `NOT NULL` this is fine.
I learned something with this, Thank you!

koenaad commented:
Okay, changes in the last few commits:
- resolved the code review comments from @jyn514 
- moved the blacklist subcomand, usage is now: `cratesfyi database blacklist add <CRATE>`
- added a subcommand `cratesfyi database blacklist list`
- `RustwideBuilder::build_package` will log an info message and return `Ok(false)` if a crate is blacklisted

As example, if `regex` is blacklisted:

```
$ docker-compose run web build crate regex 1.3.1
...
2019/12/29 16:45:46 [INFO] cratesfyi::docbuilder::rustwide_builder: found rustc rustc 1.42.0-nightly (3a3f4a7cb 2019-12-28)
2019/12/29 16:45:46 [INFO] cratesfyi::docbuilder::rustwide_builder: building package regex 1.3.1
2019/12/29 16:45:46 [INFO] cratesfyi::docbuilder::rustwide_builder: skipping build of regex, crate has been blacklisted
2019/12/29 16:45:46 [DEBUG] cratesfyi::docbuilder: Saving cache
```

Possible future additions:
- When adding a crate to the blacklist, verify if it exists in the `crates` table --> if not, the user might have made a typo.
Add force option to `blacklist add` to override this check.

pietroalbini commented:
This looks great! Huge thanks @koenaad for implementing this!

@jyn514 feel free to merge if you don't have any more blockers yourself.
